### PR TITLE
Issues/sling 11741

### PR DESCRIPTION
**Pull request by cziegeler**

### CODE REVIEW

File: src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.osgi.service.cm.ConfigurationEvent;
+import org.osgi.service.cm.ConfigurationListener;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Component(service = {ConfigurationListener.class, ConfigurationUpdater.class})
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java:42]: Can you please provide a short comment what this class is all about?

File: src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            final Dictionary<String, Object> targetProps = new Hashtable<>();
+            for(final String name : Collections.list(sourceProps.keys())) {
+                targetProps.put(this.propsToReplace.getOrDefault(name, name), sourceProps.get(name));
+            }
+            try {
+                final Configuration cfg = this.createConfiguration(sourceConfig.getPid());
+                if (cfg==null) return;
+                logger.info("Creating new configuration with PID {} for source PID: {}", cfg.getPid(), sourceConfig.getPid());
+                cfg.update(targetProps);
+                logger.info("Deleting source configuration wuth PID {} after it was migrated", sourceConfig.getPid());
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java:128]: I see 2 log statements for each update being performed. Suggestions:
* Can we merge it into a single statement?
* We should point that the developer/admin should change the configuration accordingly, maybe with a link to documentation or a JIRA ticket?

File: src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+    private class PidConfigurationUpdater extends Updater {
+
+        public PidConfigurationUpdater(final String oldPid, final String newPid, final Map<String, String> propsToReplace) {
+            super(oldPid, newPid, propsToReplace);
+        }
+
+        @Override
+        protected void updateProps(final ConfigurationEvent event) {
+            if (this.oldPid.equals(event.getPid())) {
```
Comment (sagarmiglani) [inline src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java:150]: Would it make sense to log a warning and instruct the user to use the new configuration? Also what if a user has both old and new configuration in place?

Comment (sagarmiglani) [reply src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java:150]: > Good points. We log the situation when the configuration is actually converted with a link to the docs. Do you think we need more?

I think link to the docs in enough, but logging warning instead of info seems to be a better option. "Info" messages can typically be disregarded as they provide general information.

> If the new configuration exists, it will be overwritten by the old one and then the old one will be deleted.
We could handle this differently and not touch the new configuration but just log and delete the old one?

While both solutions appear to be viable options, IMO, it makes slightly more sense not to touch the new configuration. This approach aligns with our goal of encouraging users to migrate to the newer configurations. What do you think?

File: src/main/java/org/apache/sling/jcr/base/internal/AllowListFragment.java [hunk 4]
```diff
@@ -32,59 +32,58 @@
 import static java.util.Arrays.asList;
 
 @ObjectClassDefinition(
-        name = "Apache Sling Login Admin Whitelist Configuration Fragment",
-        description = "Whitelist configuration fragments contribute a list of whitelisted bundle symbolic " +
-                "names to the Login Admin Whitelist. This allows for modularisation of the whitelist."
+        name = "Apache Sling Login Admin Allow List Configuration Fragment",
+        description = "Allow list configuration fragments contribute a list of allow listed bundle symbolic " +
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/jcr/base/internal/AllowListFragment.java:36]: I think rewording the description would make it easier to understand, just replacing "white list" with "allow list" makes it hard to read IMHO, I suggest:

> This list of Bundle Symbolic Names is added to the list of bundles which are allowed to use Administrative Login. The full list is built, in a modular way, out of all such configuration fragments.

File: src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is updating configuration and configuration property names to use
+ * more inclusive language.
+ * See https://issues.apache.org/jira/browse/SLING-11741
+ */
+@Component(service = {ConfigurationListener.class, ConfigurationUpdater.class})
+public class ConfigurationUpdater implements ConfigurationListener {
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java:48]: IIUC this updater is specific to this configuration name change, aren't we going to have more similar cases of configuration renaming? And if yes, shouldn't that class rather be generic in its own module?

We can always make it more generic later, but I'm curious about the intention.

Comment (cziegeler) [reply src/main/java/org/apache/sling/jcr/base/internal/ConfigurationUpdater.java:48]: We hope that we don't have more of those. So for now no need to have a generic solution.

**Review by jsedding — COMMENTED** [review]
To be honest, I am a bit concerned about modifying the configurations. Instead, I would prefer an implementation via `ConfigurationPlugin`, that changes the configurations on the fly.

In my opinion, configuration updates happen via deployments in most cases. If we change the configuration object in the running system, it does no longer match the deployment, which makes it harder to understand and analyse. Furthermore, we cause configuration events, which may cause unnecessary jitter in the system.

I think these issues can be avoided by using a `ConfigurationPlugin` instead. The only downside, AFAICT, is that we cannot change the old PIDs. We would likely need to allow our affected components bind the old and the new configuration PIDs.

**Comment by cziegeler** [comment]
@jsedding I'll have a look to see what we can do at runtime *without* creating/deleting configurations

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java [hunk 6]
```diff
@@ -83,23 +75,22 @@ public class LoginAdminAllowList {
...
         LOG.info("AllowListFragment removed '{}'", fragment);
     }
 
-    @Activate @Modified @SuppressWarnings("unused")
-    void configure(LoginAdminAllowListConfiguration configuration, Map<String, Object> properties) {
-        this.config = new ConfigurationState(configuration);
-        ensureBackwardsCompatibility(properties, PROP_WHITELIST_BUNDLES_DEFAULT);
-        ensureBackwardsCompatibility(properties, PROP_WHITELIST_BUNDLES_ADDITIONAL);
+    @Activate @Modified
+    void configure(final LoginAdminAllowListConfiguration configuration, final Map<String, Object> properties) {
```
Comment (jsedding) [inline src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:90]: What about using a second annotation, without metatype, like you did with the [`VanityPathConfigurer.DeprecatedVanityConfig`](https://github.com/apache/sling-org-apache-sling-resourceresolver/blob/a24a17467b334f9076f91c2dccd2491dfc9ae96b/src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverFactoryActivator.java#L221)?

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java [hunk 7]
```diff
@@ -133,24 +124,47 @@ public boolean allowLoginAdministrative(Bundle b) {
...
 
-        private ConfigurationState(final LoginAdminAllowListConfiguration config) {
             final String regexp = config.allowlist_bundles_regexp();
-            if(regexp.trim().length() > 0) {
-                allowListRegexp = Pattern.compile(regexp);
-                LOG.warn("A 'allowlist.bundles.regexp' is configured, this is NOT RECOMMENDED for production: {}",
-                        allowListRegexp);
+            if (regexp.trim().length() > 0) {
+                if (legacyRegexp != null) {
+                    LOG.warn("Both deprecated configuration property '{}' and configuration property '{}' are set. " +
```
Comment (jsedding) [inline src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:155]: IMHO it could be clearer that the second property name is not deprecated. Maybe reword to something like "... and its non-deprecated equivalent '{}' are set.". WDYT?

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java [hunk 8]
```diff
@@ -133,24 +124,47 @@ public boolean allowLoginAdministrative(Bundle b) {
...
+                this.allowListRegexp = Pattern.compile(regexp);
             } else {
-                allowListRegexp = null;
+                this.allowListRegexp = legacyRegexp != null ? Pattern.compile(legacyRegexp) : null;
             }
-
-            bypassAllowList = config.allowlist_bypass();
-            if(bypassAllowList) {
+            if (this.allowListRegexp != null) {
+                LOG.warn("A 'allowlist.bundles.regexp' is configured, this is NOT RECOMMENDED for production: {}", allowListRegexp);
```
Comment (jsedding) [inline src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:164]: Should we deprecate the regexp for removal? And only leave support for "whitelist.bundles.regexp"?

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java [hunk 9]
```diff
@@ -133,24 +124,47 @@ public boolean allowLoginAdministrative(Bundle b) {
...
+            }
+            this.bypassAllowList = bypass;
+            if (this.bypassAllowList) {
                 LOG.info("bypassAllowlist=true, allowlisted BSNs=<ALL>");
                 LOG.warn("All bundles are allowed to use loginAdministrative due to the 'allowlist.bypass' " +
```
Comment (jsedding) [inline src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:169]: Deprecate for removal? Or is it too useful for development?

**Review by jsedding — APPROVED** [review]
I made some suggestions and added a few questions around whether we should not re-introduce the "regexp" and "bypass" config flags in the new configs. Feel free to merge if you think we should keep them. All other changes could also be done in a new PR, I think.

File: src/main/java/org/apache/sling/jcr/base/internal/LegacyFragment.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    @Activate
+    public LegacyFragment(final @Reference LoginAdminAllowList allowList, final Map<String, Object> config) {
+        LoginAdminAllowList.LOG.warn("Using deprecated factory configuration '{}'. " +
+            "Update your configuration to use configuration '{}' instead.", 
+            LEGACY_FACTORY_PID, AllowListFragment.FACTORY_PID);
+        this.allowList = allowList;
+        final String name = Converters.standardConverter().convert(config.get(LEGACY_NAME)).to(String.class);
+        final String[] bundles = Converters.standardConverter().convert(config.get(LEGACY_BUNDLES)).to(String[].class);
+        this.fragment = new AllowListFragment(name, bundles);
+        this.allowList.bindAllowListFragment(fragment);
```
[the remark this answers is not in the document]
Comment (cziegeler) [reply src/main/java/org/apache/sling/jcr/base/internal/LegacyFragment.java:58]: Yes, I decided against this as it is an anti-pattern to register services within an activate method.

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java (hunk 6 above)
Comment (cziegeler) [reply src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:90]: We could but we need the properties for all the methods below, so it is a little bit easier this way

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java (hunk 7 above)
Comment (cziegeler) [reply src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:155]: Please suggest a better wording :)

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java [hunk 11]
```diff
@@ -47,34 +45,28 @@
...
-        @Reference(
-            name = "ConfigurationUpdater",
-            service = ConfigurationUpdater.class,
-            cardinality = ReferenceCardinality.MANDATORY
-        )
-    }
-)
-@Designate(
-        ocd = LoginAdminAllowListConfiguration.class
+    configurationPid = {LoginAdminAllowList.PID, LoginAdminAllowList.LEGACY_PID}
```
[the remark this answers is not in the document]
Comment (cziegeler) [reply src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:48]: As there is no overlap in properties I think it does not matter

[the remark this answers is not in the document]
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:48]: That's right, yes.

File: src/main/java/org/apache/sling/jcr/base/internal/LegacyFragment.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/base/internal/LegacyFragment.java:58]: I din't mean to register the service manually. I meant to have different implementations of the `AllowListFragment`, both DS components registered as services.

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java (hunk 7 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java]: Fixed.

File: src/main/java/org/apache/sling/jcr/base/internal/LegacyFragment.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/base/internal/LegacyFragment.java]: See ce1e6a9 for what I mean.

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java (hunk 8 above)
Comment (cziegeler) [reply src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:167]: if we do this we should do it in a different issue

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java (hunk 9 above)
Comment (cziegeler) [reply src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:172]: Same as above

File: src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java (hunk 8 above)
Comment (jsedding) [reply src/main/java/org/apache/sling/jcr/base/internal/LoginAdminAllowList.java:167]: Fine with me.

**Comment by jsedding** [comment]
@cziegeler if you're happy with the current state, feel free to merge!
